Validate invalid base type declarations in ILVerify (#129118)
dotnet/runtime: docs/design/specs/Ecma-335-Augments.md

Reviewer: ```suggestion
* https://github.com/dotnet/runtime/issues/4945
* https://github.com/dotnet/runtime/pull/123819#discussion_r2810397721
```

3. PEVerify, the CLR runtime and C# compiler prior to VS 2015 report an error when encountering an encoded TypeSpec in the positions described above.

### 2. `(CMOD_OPT | CMOD_REQ) <TypeSpec>` is invalid

In II.23.2.7, it is noted that CMOD_OPT or CMOD_REQD is followed
by a TypeRef or TypeDef metadata token. TypeSpec tokens in this
position should be treated as invalid metadata.

> Allowing TypeSpec custom modifier types also creates a path for

Reviewer: ```suggestion
Allowing TypeSpec custom modifier types creates a path for
```

recursive TypeSpec encodings, which has caused problems in runtimes

> and tools. Such metadata should be rejected rather than cycle-detected.

Reviewer: ```suggestion
recursive TypeSpec encodings that are complicated to detect. Such metadata should be rejected rather than cycle-detected.
```

3. PEVerify, the CLR runtime and C# compiler prior to VS 2015 report an error when encountering an encoded TypeSpec in the positions described above.

### 2. `(CMOD_OPT | CMOD_REQ) <TypeSpec>` is invalid

In II.23.2.7, it is noted that CMOD_OPT or CMOD_REQD is followed
by a TypeRef or TypeDef metadata token. TypeSpec tokens in this
position should be treated as invalid metadata.

> Allowing TypeSpec custom modifier types creates a path for

Reviewer: Nit: Could you please reformat this to have similar text width, and add new line before "Related issues"

Author: Sure, I’ve updated it.

with

> The CMOD_OPT or CMOD_REQD is followed by a metadata token that indexes a row in the TypeDef

> table or TypeRef table. However, these tokens are encoded and compressed –

Reviewer: This is not proposing any actual change anymore. Both original and proposed replacements are identical. If we want to propose any change here, it should be to add a sentence that explains why typespecs are not allowed.

Author: Updated. I added a sentence explaining that TypeSpec tokens are not valid in this position because allowing them could form recursive TypeSpec encodings.
